qiskit/circuit/quantumcircuit.py

```diff
@@ -1419,30 +1384,18 @@ def append(
         expanded_qargs = [self.qbit_argument_conversion(qarg) for qarg in qargs or []]
         expanded_cargs = [self.cbit_argument_conversion(carg) for carg in cargs or []]
 
-        if self._control_flow_scopes:
-            circuit_data = self._control_flow_scopes[-1].instructions
-            appender = self._control_flow_scopes[-1].append
-            requester = self._control_flow_scopes[-1].request_classical_resource
-        else:
-            circuit_data = self._data
-            appender = self._append
-            requester = self._resolve_classical_resource
-        instructions = InstructionSet(resource_requester=requester)
-        if isinstance(operation, Instruction):
-            for qarg, carg in operation.broadcast_arguments(expanded_qargs, expanded_cargs):
-                self._check_dups(qarg)
-                data_idx = len(circuit_data)
-                appender(CircuitInstruction(operation, qarg, carg))
-                instructions._add_ref(circuit_data, data_idx)
-        else:
-            # For Operations that are non-Instructions, we use the Instruction's default method
-            for qarg, carg in Instruction.broadcast_arguments(
-                operation, expanded_qargs, expanded_cargs
-            ):
-                self._check_dups(qarg)
-                data_idx = len(circuit_data)
-                appender(CircuitInstruction(operation, qarg, carg))
-                instructions._add_ref(circuit_data, data_idx)
+        instructions = InstructionSet(resource_requester=api.resolve_classical_resource)
+        # For Operations that are non-Instructions, we use the Instruction's default method
+        broadcast_iter = (
+            operation.broadcast_arguments(expanded_qargs, expanded_cargs)
+            if isinstance(operation, Instruction)
+            else Instruction.broadcast_arguments(operation, expanded_qargs, expanded_cargs)
+        )
+        for qarg, carg in broadcast_iter:
+            self._check_dups(qarg)
+            instruction = CircuitInstruction(operation, qarg, carg)
+            api.append(instruction)
+            instructions._add_ref(api.instructions, len(api.instructions) - 1)
```

Comment:
The `broadcast_iter` is the only unrelated change. Originally this was only collapsing `appender` into `api.append` (since we already have the `api` object), but since #10827 the `api.instructions` field is dispatched as well, so it was growing to quite a few dispatched names.